rosidl_parser/rosidl_parser/definition.py

```diff
@@ -0,0 +1,343 @@
+# Copyright 2018 Open Source Robotics Foundation, Inc.
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+from collections import OrderedDict
+import pathlib
+
+
+BASIC_TYPES = [
+    'short',
+    'unsigned short',
+    'long',
+    'unsigned long',
+    'long long',
+    'unsigned long long',
+    'float',
+    'double',
+    'long double',
+    'char',
+    'wchar',
+    'boolean',
+    'octet',
+    'int8',
+    'uint8',
+    'int16',
+    'uint16',
+    'int32',
+    'uint32',
+    'int64',
+    'uint64',
+]
+
+SERVICE_REQUEST_MESSAGE_SUFFIX = '_Request'
+SERVICE_RESPONSE_MESSAGE_SUFFIX = '_Response'
+
+
+class AbstractType:
+
+    __slots__ = ()
+
+    def __eq__(self, other):
+        return type(self) == type(other)
+
+
+class BaseType(AbstractType):
```

Comment:
`AbstractType` is the root type. All non-nested types inherit from `BaseType` but arrays and sequences do not.

Reply:
Why don't arrays and sequences inherit from `BaseType`? Is it because there is a place where arrays and sequences are forbidden, but all other types are allowed?

Reply:
> Is it because there is a place where arrays and sequences are forbidden ...

Yes, an array or sequence con only contain `BaseType` elements.